include/box.hpp

```diff
@@ -26,7 +26,7 @@ class box {
   }
 
   /// Compute ray interaction with the box
-  bool hit(auto& ctx, const ray& r, real_t min, real_t max, hit_record& rec,
+  bool hit(const ray& r, real_t min, real_t max, hit_record& rec,
            material_t& hit_material_type) const {
     hit_record temp_rec;
     material_t temp_material_type;
```

Comment:
I have the feeling that if we remove the `std::monostate` from the material it will always create a `sphere` here even when there is no intersection.